File: archive/RTCFileVideoCapturer+URL.m
//
//  RTCFileVideoCapturer+URL.m
//  WebRTCDemo
//
//  Implementation of URL support for RTCFileVideoCapturer
//

#import "RTCFileVideoCapturer+URL.h"
#import <AVFoundation/AVFoundation.h>
#import <objc/runtime.h>

// Define the status enum to match the private implementation
typedef NS_ENUM(NSInteger, RTCFileVideoCapturerStatus) {
    RTCFileVideoCapturerStatusNotInitialized,
    RTCFileVideoCapturerStatusStarted,
    RTCFileVideoCapturerStatusStopped
};

@interface RTCFileVideoCapturer (Private)
@property(nonatomic, strong) NSURL *fileURL;
- (void)setupReaderOnError:(id)errorBlock;
@end

@implementation RTCFileVideoCapturer (URL)

- (void)startCapturinggggggFromFileURL:(NSURL *)fileURL
                          onError:(nullable RTCFileVideoCapturerURLErrorBlock)errorBlock {
    NSLog(@"RTCFileVideoCapturer: Starting capture from file URL: %@", fileURL);
    
    // Set the status to started (access the private _status ivar)
    Ivar statusIvar = class_getInstanceVariable([self class], "_status");
    if (statusIvar) {
        NSInteger status = RTCFileVideoCapturerStatusStarted;
        // Set the ivar value directly using pointer manipulation (safe for primitive types)
        ptrdiff_t offset = ivar_getOffset(statusIvar);
        *(NSInteger *)((char *)(__bridge void *)self + offset) = status;
    }
    
    // Dispatch to background queue like the original implementation
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        // Set the presentation time to zero
        CMTime zeroTime = CMTimeMake(0, 0);
        NSValue *timeValue = [NSValue valueWithCMTime:zeroTime];
        [self setValue:timeValue forKey:@"lastPresentationTime"];
        
        // Set the file URL using the private property
        [self setValue:fileURL forKey:@"fileURL"];
        
        // Call the private setup method
        if ([self respondsToSelector:@selector(setupReaderOnError:)]) {
            NSMethodSignature *signature = [self methodSignatureForSelector:@selector(setupReaderOnError:)];
            NSInvocation *invocation = [NSInvocation invocationWithMethodSignature:signature];
            [invocation setTarget:self];
            [invocation setSelector:@selector(setupReaderOnError:)];
            
            // Pass the error block as parameter
            id block = errorBlock ? errorBlock : ^(NSError *error) {
                NSLog(@"RTCFileVideoCapturer error: %@", error);
            };
            [invocation setArgument:&block atIndex:2]; // index 0 is self, 1 is _cmd, 2 is first arg
            [invocation invoke];
        } else {
            if (errorBlock) {
                NSError *error = [NSError errorWithDomain:@"RTCFileVideoCapturer"
                                                     code:-1
                                                 userInfo:@{NSLocalizedDescriptionKey: @"setupReaderOnError method not available"}];
                errorBlock(error);
            }
        }
    });
}

@end

File: ios/WebRTCDemo/RTCFileVideoCapturer+URL.swift
//
//  RTCFileVideoCapturer+URL.swift
//  WebRTCDemo
//
//  Extension to add support for capturing from arbitrary file URLs
//  RTCFileVideoCapturer only exposes a method to capture from bundle resources. E.g: fileVideoCapturer.startCapturing(fromFileNamed: "video.mp4")
//  We need to extend it to support arbitrary file URLs.

import Foundation
import WebRTC

extension RTCFileVideoCapturer {
    
    /// Start capturing from a file at the given URL
    /// - Parameters:
    ///   - fileURL: The URL of the video file to capture from
    ///   - onError: Error callback
    func startCapturing(fromFileURL fileURL: URL, onError: @escaping (Error) -> Void) {
        print("Starting RTCFileVideoCapturer with file URL: \(fileURL)")
        // Dispatch to background queue (mimicking the original implementation)
        DispatchQueue.global(qos: .default).async { [weak self] in
            guard let self = self else { return }
            
            // Set status to started (1)
            self.setValue(1, forKey: "_status")
            
            // Reset presentation time
            self.setValue(CMTime.zero, forKey: "_lastPresentationTime")
            
            // Set the file URL
            self.setValue(fileURL, forKey: "_fileURL")
            
            // Call the private setupReaderOnError: method
            let selector = NSSelectorFromString("setupReaderOnError:")
            if self.responds(to: selector) {
                // Create error block wrapper
                let errorBlock: @convention(block) (Error) -> Void = { error in
                    onError(error)
                }
                
                // Call the method
                self.perform(selector, with: errorBlock)
            } else {
                print("RTCFileVideoCapturer does not respond to setupReaderOnError:")
            }
        }
    }
}
